component/stream/stream.go

```diff
@@ -163,117 +128,40 @@ func (st *Stream) unlockBlock(block *cache_policy.CacheBlock, exists bool) {
 }
 
-func (st *Stream) getBlock(handle *handlemap.Handle, offset int64) (*cache_policy.CacheBlock, bool, error) {
-	blockSize := st.cache.blockSize
-	blockKeyObj := offset
-	handle.CacheObj.Lock()
-	block, found := handle.CacheObj.Get(blockKeyObj)
-	if !found {
-		if (offset + blockSize) > handle.Size {
-			blockSize = handle.Size - offset
-		}
-		block = &cache_policy.CacheBlock{
-			StartIndex: offset,
-			EndIndex:   offset + blockSize,
-			Data:       make([]byte, blockSize),
-			Last:       (offset + blockSize) >= handle.Size,
-		}
-		block.Lock()
-		handle.CacheObj.Put(blockKeyObj, block)
-		handle.CacheObj.Unlock()
-		// if the block does not exist fetch it from the next component
-		options := internal.ReadInBufferOptions{
-			Handle: handle,
-			Offset: block.StartIndex,
-			Data:   block.Data,
-		}
-		_, err := st.NextComponent().ReadInBuffer(options)
-		if err != nil && err != io.EOF {
-			return nil, false, err
-		}
-		return block, false, nil
-	} else {
-		block.RLock()
-		handle.CacheObj.Unlock()
-		return block, true, nil
-	}
-}
-
-func (st *Stream) copyCachedBlock(handle *handlemap.Handle, offset int64, data []byte) (int, error) {
-	dataLeft := int64(len(data))
-	// counter to track how much we have copied into our request buffer thus far
-	dataRead := 0
-	// covers the case if we get a call that is bigger than the file size
-	for dataLeft > 0 && offset < handle.Size {
-		// round all offsets to the specific blocksize offsets
-		cachedBlockStartIndex := (offset - (offset % st.cache.blockSize))
-		// Lock on requested block and fileName to ensure it is not being rerequested or manipulated
-		block, exists, err := st.getBlock(handle, cachedBlockStartIndex)
-		if err != nil {
-			st.unlockBlock(block, exists)
-			log.Err("Stream::ReadInBuffer : failed to download block of %s with offset %d: [%s]", handle.Path, block.StartIndex, err.Error())
-			return dataRead, err
-		}
-		dataCopied := int64(copy(data[dataRead:], block.Data[offset-cachedBlockStartIndex:]))
-		st.unlockBlock(block, exists)
-		dataLeft -= dataCopied
-		offset += dataCopied
-		dataRead += int(dataCopied)
-	}
-	return dataRead, nil
+func (st *Stream) OpenFile(options internal.OpenFileOptions) (*handlemap.Handle, error) {
+	return st.cache.OpenFile(options)
 }
 
 func (st *Stream) ReadInBuffer(options internal.ReadInBufferOptions) (int, error) {
-	// if we're only streaming then avoid using the cache
-	if st.streamOnly {
-		data, err := st.NextComponent().ReadInBuffer(options)
-		if err != nil && err != io.EOF {
-			log.Err("Stream::ReadInBuffer : error failed to download requested data for %s: [%s]", options.Handle.Path, err.Error())
-		}
-		return data, err
-	}
-	return st.copyCachedBlock(options.Handle, options.Offset, options.Data)
+	return st.cache.ReadInBuffer(options)
 }
 
 func (st *Stream) WriteFile(options internal.WriteFileOptions) (int, error) {
-	return st.NextComponent().WriteFile(options)
+	return st.cache.WriteFile(options)
```

Comment:
If we are in streaming only mode, then we need to forward the write call to next component?

Reply:
yes, if its stream only mode it will go to next comp